%R CMD Sweave --pdf HaplotypR.Rnw
%\VignetteIndexEntry{An introduction to HaplotypR}
%\VignetteDepends{}
%\VignetteKeywords{XXXKexword}
%\VignettePackage{HaplotypR}

\documentclass[12pt]{article}

<<style, eval=TRUE, echo=FALSE, results=tex>>=
BiocStyle::latex(width=80, bibstyle="unsrturl")
@

%\usepackage{cite} % handled by BiocStyle
\usepackage{Sweave}
%\usepackage{times}
\usepackage{inconsolata}
%\usepackage[scaled=0.85]{beramono}
\usepackage{url}
%\usepackage[breaklinks]{hyperref} % remove to suppress links % already loaded in BiocStyle
%\usepackage[utf8x]{inputenc}
%\usepackage[authoryear,round]{natbib}


%\newcommand{\Rfunction}[1]{{\texttt{#1}}} % defined in BiocStyle
%\newcommand{\Robject}[1]{{\texttt{#1}}}   % defined in BiocStyle
%\newcommand{\Rpackage}[1]{{\textit{#1}}}  % replaced by \Biocpkg, \Biocannopkg, \Biocexptpkg, \CRANpkg and \Rpackagefrom BiocStyle
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rfunarg}[1]{{\texttt{#1}}}
%\newcommand{\Rclass}[1]{{\textit{#1}}}    % defined in BiocStyle
%\newcommand{\Rcode}[1]{{\texttt{#1}}}     % defined in BiocStyle

%\newcommand{\software}[1]{\textsf{#1}} % defined in BiocStyle
%\newcommand{\R}{\software{R}}          % defined in BiocStyle
%\newcommand{\HaplotypR}{\Rpackage{HaplotypR}}  % adjusted to BiocStyle
%\newcommand{\HaplotypR}{\Biocpkg{HaplotypR}}
\newcommand{\fasta}{\texttt{FASTA}}
\newcommand{\fastq}{\texttt{FASTQ}}

%original code environment
%\DefineVerbatimEnvironment{Sinput}{Verbatim}{fontshape=sl}
%\DefineVerbatimEnvironment{Soutput}{Verbatim}{}
%\DefineVerbatimEnvironment{Scode}{Verbatim}{fontshape=sl}

%code environment used before BiocStyle
%\usepackage{Sweave}
%\DefineVerbatimEnvironment{Sinput}{Verbatim} {xleftmargin=2em}
%\DefineVerbatimEnvironment{Soutput}{Verbatim}{xleftmargin=2em}
%\DefineVerbatimEnvironment{Scode}{Verbatim}{xleftmargin=2em}
%\fvset{listparameters={\setlength{\topsep}{0pt}}}
%\renewenvironment{Schunk}{\vspace{\topsep}}{\vspace{\topsep}}

%\bioctitle{An Introduction to \HaplotypR{}}
\bioctitle{An Introduction to HaplotypR}
\author{Anita Lerch}
\date{Modified: December 19 2014. Compiled: \today}

\begin{document}
\SweaveOpts{concordance=TRUE}
%\bibliographystyle{plain}
%\bibliographystyle{unsrt} % handled by BiocStyle
%\bibliographystyle{plainnat}

<<options, results=hide, echo=FALSE>>=
#options(width=65)
options('useFancyQuotes' = FALSE, continue=" ", digits=3)
@


\maketitle

\tableofcontents

\newpage


\section{Introduction}

% The \Biocpkg{HaplotypR} package integrates the functionality of several \R{} packages (such as \Biocpkg{Rswarm} \cite{swarm} and \Biocpkg{Rvsearch} \cite{swarm}) and external software (e.g. \software{bowtie}, through the \Biocpkg{Rbowtie} package). The package aims to .


\section{Preliminaries}

\subsection{Citing \Biocpkg{HaplotypR}}
If you use \Biocpkg{HaplotypR} \cite{HaplotypR} in your work, you can cite it as follows:
<<cite, eval=TRUE>>=
citation("HaplotypR")
@

\subsection{Installation}
\label{sec:Installation}
# <<install, eval=FALSE>>=
# source("http://www.bioconductor.org/biocLite.R")
# biocLite("HaplotypR")
# @


\newpage
\section{Quick Start}


Before you can use any \Biocpkg{HaplotypR} commands you have to load the package by typing
<<loadHaplotypRLibrary, eval=FALSE>>=
library(HaplotypR)
@ 

\subsection{Sample \Biocpkg{HaplotypR} session}
\label{sec:SampleSession}
<<SampleSession1, eval=TRUE>>=
library(HaplotypR)
file.copy(system.file(package="HaplotypR", "extdata"), ".", recursive=TRUE)
@


\section{Session information}
The output in this vignette was produced under:
<<sessionInfo, results=tex, echo=FALSE>>=
toLatex(sessionInfo())
@

<<cleanUp, eval=TRUE, echo=FALSE>>=
unlink("extdata", recursive=TRUE, force=TRUE)
@

\bibliography{HaplotypR-refs}

\end{document}



% LocalWords:  Biostrings nBases minLength outfiles
